### ISSUE TITLE
Define `@cucumber/cucumber` as a peer dependency, to avoid problems with importing its commands (when, ...)

### ISSUE THREAD

**Issue by dietergeerts** [webdriverio/webdriverio#7234]
**Is your feature request related to a problem? Please describe.**
When I do not add "@cucumber/cucumber" to my own dependencies, I am unable to import its commands. But when I do add it, I have another version than what this package is using, which can lead to problems in our end-2-end tests.

**Describe the solution you'd like**
Define "@cucumber/cucumber" as a peer dependency

**Describe alternatives you've considered**
Re-export the cucumber commands, but that's not clean Imho.

**Additional context**
I use TypeScript, and am transpiling with babel, not ts-node, through cucumber options.

**Comment by praveendvd** [comment]
```
  "devDependencies": {
    "@rpii/wdio-html-reporter": "^7.7.1",
    "@types/yargs": "^17.0.2",
    "@wdio/allure-reporter": "^7.8.0",
    "@wdio/cucumber-framework": "^7.8.0",
    "@wdio/dot-reporter": "^7.8.0",
    "@wdio/junit-reporter": "^7.8.0",
    "@wdio/local-runner": "^7.8.0",
    "@wdio/spec-reporter": "^7.8.0",
    "allure-commandline": "^2.13.8",
    "chromedriver": "^91.0.1",
    "eslint": "^7.31.0",
    "eslint-config-google": "^0.14.0",
    "wdio-chromedriver-service": "^7.2.0",
    "wdio-json-reporter": "^2.0.0"
  }
```

this is my dependencies and @cucumber is available by default in the node modules  as its a dev dependency for cucumber framework


https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-cucumber-framework/package.json

@christian-bromann  what would be required here , i think its already available ?

**Comment by christian-bromann** [comment]
@praveendvd I believe the issue is if you want to use Cucumber for other testing purposes and you want to add it to your own project dependencies than it would clash with the ones required by `@wdio/cucumber-framework`. However afaik moving it to the  `peerDependency` section would mean that this dependency is not automatically installed if you use NPM v3 - v6 (see [release post](https://blog.npmjs.org/post/110924823920/npm-weekly-5)). In v7 this behavior was added again (see [rfc](https://github.com/npm/rfcs/blob/latest/implemented/0025-install-peer-deps.md)). So I am honestly not quite sure what the best approach is.

@dietergeerts can you maybe give some more insights how this causes issues in your e2e tests?

**Comment by dietergeerts** [comment]
I had several issues trying to get my end-2-end tests to work, and was able to now pinpoint it to https://github.com/testing-library/webdriverio-testing-library/issues/29. So currently, this dependency doesn't give me issues. 

That said, there is no reason to have this as an actual dependency AND it potentially will bring some problems because when you use cucumber, you'll import it's step factories, meaning that they will come from the version you have stated in your own dependencies.

```
import { Given, When } from '@cucumber/cucumber'
import { authLogin } from './auth-login.fragment'

Given(/^I have logged in into the application$/, async () => {
  await authLogin.login()
})
When(/^I navigate to the Demetra application$/, async () => {
  await authLogin.navigate()
  await authLogin.waitForLoaded()
})
```
So therefore, it would be better to avoid future problems and state it as a peerDependency instead.

In the past, I have encountered several other problems with situations like this, packages that have a dependency while it is actually a peer dependency, and the versions are in conflict, and not always easy to resolve. Luckily, I use pnpm where I can manipulate the dependencies upon install, but still it's a problem with the package that must be solved.

**Comment by BorisOsipov** [comment]
`@wdio/cucumber-framework` uses cucumber package as main dependency. It cannot be peer dependency under any circumstances.

**Comment by dietergeerts** [comment]
@BorisOsipov, I see. Then another solution would be to re-export its exported values, to make sure we use the ones from the version used to run the cucumber part.

**Comment by BorisOsipov** [comment]
I am not sure that re-export its exported values is a good idea.

>not defining the @cucumber/cucumber package as dependency in our own project is problematic, because of several tooling (TS, ....) and linting expect that we define what we use.)

I don't know why it is problematic. You can add cucumber package as dependency for sake of linters. Cucumber follows semver - you should not run into api incompatibility issues.

**Comment by christian-bromann** [comment]
@dietergeerts I see the issue here. Let's say you use Cucumber for some other testing practices then and that version differs from the one used in WebdriverIO then there could be compatibility issues because the Cucumber version that is defined in the projects package.json will always be preferred.

@BorisOsipov is totally right here, we can't make this a peer dependency because `@wdio/cucumber-framework` relies on a specific version of Cucumber.

My suggestion therefor is to re-export these properties in rare cases where someone uses a different Cucumber version, e.g.:

```js
import { Given, When, Then } from '@wdio/cucumber-framework`
```

Any help on this would be appreciated!

**Comment by dietergeerts** [comment]
@christian-bromann , yup, that's the exact problem. This way, there can be 2 instances being used, which can lead to strange bugs that are hard to resolve. I say "can", because it depends on the package. So for making sure we don't need to mind, it has to be a peer-dependency OR using re-exports, as then it's like a replacement for it. I'll see what I can do to implement this. The import example you stated is one to go for?

**Comment by christian-bromann** [comment]
> The import example you stated is one to go for?

Yes, I would say it is the best approach. Again, we can't move Cucumber to be a peer dependency as we can't ensure it will be automatically downloaded and we need to rely on a specific Cucumber version. Re-exporting feels most naturally to me.

**Comment by christian-bromann** [comment]
implemented in #7398
